File: azure/producer/transaction_producer.py
#!/usr/bin/env python3
"""
Azure Event Hubs Transaction Producer
=====================================
Envoie des transactions simulées vers Azure Event Hubs (Kafka-compatible)

Usage:
    python transaction_producer.py --count 10000 --fraud-rate 0.03

Prérequis:
    pip install azure-eventhub python-dotenv
"""

import os
import json
import time
import random
import argparse
from datetime import datetime
from typing import List, Dict
from dotenv import load_dotenv

from azure.eventhub import EventHubProducerClient, EventData

# Charger les variables d'environnement
load_dotenv()


class TransactionGenerator:
    """Génère des transactions bancaires réalistes"""

    LOCATIONS = [
        {"city": "Paris", "country": "FR", "timezone": "Europe/Paris"},
        {"city": "London", "country": "GB", "timezone": "Europe/London"},
        {"city": "New York", "country": "US", "timezone": "America/New_York"},
        {"city": "Tokyo", "country": "JP", "timezone": "Asia/Tokyo"},
        {"city": "Singapore", "country": "SG", "timezone": "Asia/Singapore"},
        {"city": "Dubai", "country": "AE", "timezone": "Asia/Dubai"},
        {"city": "Sydney", "country": "AU", "timezone": "Australia/Sydney"},
        {"city": "Berlin", "country": "DE", "timezone": "Europe/Berlin"},
    ]

    MERCHANT_CATEGORIES = [
        {"code": "GROCERY", "avg_amount": 85, "std": 45, "fraud_risk": 0.01},
        {"code": "RESTAURANT", "avg_amount": 45, "std": 30, "fraud_risk": 0.02},
        {"code": "RETAIL", "avg_amount": 150, "std": 100, "fraud_risk": 0.03},
        {"code": "ONLINE", "avg_amount": 120, "std": 80, "fraud_risk": 0.05},
        {"code": "TRAVEL", "avg_amount": 450, "std": 300, "fraud_risk": 0.04},
        {"code": "ENTERTAINMENT", "avg_amount": 65, "std": 40, "fraud_risk": 0.02},
        {"code": "UTILITIES", "avg_amount": 120, "std": 50, "fraud_risk": 0.01},
        {"code": "HEALTHCARE", "avg_amount": 200, "std": 150, "fraud_risk": 0.01},
        {"code": "GAMBLING", "avg_amount": 500, "std": 400, "fraud_risk": 0.15},
        {"code": "CRYPTO", "avg_amount": 1000, "std": 800, "fraud_risk": 0.20},
    ]

    CHANNELS = ["ONLINE", "POS", "ATM", "MOBILE", "PHONE"]
    CARD_TYPES = ["CREDIT", "DEBIT", "PREPAID"]
    CURRENCIES = ["EUR", "USD", "GBP", "JPY", "SGD"]

    def __init__(self, num_customers: int = 5000, num_merchants: int = 500):
        self.num_customers = num_customers
        self.num_merchants = num_merchants
        self.tx_counter = 0

        # Pré-générer des profils clients
        self.customer_profiles = self._generate_customer_profiles()

    def _generate_customer_profiles(self) -> Dict:
        """Génère des profils clients avec des comportements différents"""
        profiles = {}
        for i in range(self.num_customers):
            customer_id = f"CUST{i:06d}"
            profiles[customer_id] = {
                "avg_monthly_spend": random.uniform(500, 5000),
                "preferred_location": random.choice(self.LOCATIONS),
                "preferred_channel": random.choice(self.CHANNELS[:3]),
                "risk_score": random.uniform(0, 0.3),
            }
        return profiles

    def generate_normal_transaction(self) -> Dict:
        """Génère une transaction normale"""
        self.tx_counter += 1

        # Sélectionner client et marchand
        customer_id = f"CUST{random.randint(0, self.num_customers-1):06d}"
        merchant_id = f"MERCH{random.randint(0, self.num_merchants-1):05d}"

        # Sélectionner catégorie marchand
        category = random.choice(self.MERCHANT_CATEGORIES[:8])  # Exclure GAMBLING/CRYPTO pour normal

        # Générer montant basé sur la catégorie
        amount = random.gauss(category["avg_amount"], category["std"])
        amount = max(1.0, amount)  # Minimum $1

        # Localisation
        location = random.choice(self.LOCATIONS)
        customer_location = self.customer_profiles[customer_id]["preferred_location"]
        is_international = location["country"] != customer_location["country"]

        # Timestamp
        timestamp = int(time.time() * 1000)

        return {
            "transactionId": f"TX{timestamp}-{self.tx_counter:08d}",
            "customerId": customer_id,
            "merchantId": merchant_id,
            "amount": round(amount, 2),
            "currency": random.choice(self.CURRENCIES[:3]),
            "channel": random.choice(self.CHANNELS),
            "location": location["city"],
            "country": location["country"],
            "timestamp": timestamp,
            "timestampStr": datetime.utcnow().isoformat() + "Z",
            "cardType": random.choice(self.CARD_TYPES),
            "isInternational": is_international,
            "merchantCategory": category["code"],
            "previousBalance": round(random.uniform(100, 15000), 2),
            "isFraud": False,
            "fraudType": None
        }

    def generate_fraudulent_transaction(self) -> Dict:
        """Génère une transaction frauduleuse"""
        tx = self.generate_normal_transaction()
        tx["isFraud"] = True

        # Type de fraude
        fraud_type = random.choice([
            "high_amount",
            "risky_merchant",
            "velocity_abuse",
            "geo_anomaly",
            "card_testing"
        ])

        tx["fraudType"] = fraud_type

        if fraud_type == "high_amount":
            # Montant anormalement élevé
            tx["amount"] = round(random.uniform(5000, 20000), 2)

        elif fraud_type == "risky_merchant":
            # Marchand à risque
            tx["merchantCategory"] = random.choice(["GAMBLING", "CRYPTO"])
            tx["amount"] = round(random.uniform(1000, 8000), 2)

        elif fraud_type == "velocity_abuse":
            # Plusieurs transactions rapides (on simule avec un flag)
            tx["amount"] = round(random.uniform(100, 500), 2)
            tx["velocityFlag"] = True

        elif fraud_type == "geo_anomaly":
            # Transaction depuis un pays inhabituel
            tx["isInternational"] = True
            tx["country"] = random.choice(["NG", "RU", "CN", "BR"])
            tx["amount"] = round(random.uniform(500, 3000), 2)

        elif fraud_type == "card_testing":
            # Petits montants pour tester la carte
            tx["amount"] = round(random.uniform(0.50, 5.00), 2)
            tx["channel"] = "ONLINE"

        return tx

    def generate_batch(self, count: int, fraud_rate: float = 0.03) -> List[Dict]:
        """Génère un batch de transactions"""
        transactions = []
        fraud_count = 0

        for _ in range(count):
            if random.random() < fraud_rate:
                tx = self.generate_fraudulent_transaction()
                fraud_count += 1
            else:
                tx = self.generate_normal_transaction()
            transactions.append(tx)

        return transactions, fraud_count


class EventHubsProducer:
    """Producteur Azure Event Hubs"""

    def __init__(self, connection_string: str, eventhub_name: str):
        self.connection_string = connection_string
        self.eventhub_name = eventhub_name
        self.producer = None

    def connect(self):
        """Établir la connexion"""
        self.producer = EventHubProducerClient.from_connection_string(
            conn_str=self.connection_string,
            eventhub_name=self.eventhub_name
        )
        print(f"✅ Connecté à Event Hub: {self.eventhub_name}")

    def send_batch(self, transactions: List[Dict]) -> int:
        """Envoyer un batch de transactions"""
        event_data_batch = self.producer.create_batch()
        sent_count = 0

        for tx in transactions:
            try:
                # Créer l'événement avec la clé de partition
                event = EventData(json.dumps(tx))
                event.properties = {
                    "customerId": tx["customerId"],
                    "merchantCategory": tx["merchantCategory"]
                }
                event_data_batch.add(event)
                sent_count += 1
            except ValueError:
                # Batch plein, envoyer et créer nouveau
                self.producer.send_batch(event_data_batch)
                event_data_batch = self.producer.create_batch()
                event_data_batch.add(EventData(json.dumps(tx)))
                sent_count += 1

        # Envoyer le dernier batch
        if sent_count > 0:
            self.producer.send_batch(event_data_batch)

        return sent_count

    def close(self):
        """Fermer la connexion"""
        if self.producer:
            self.producer.close()
            print("✅ Connexion fermée")


def main():
    parser = argparse.ArgumentParser(description="Azure Event Hubs Transaction Producer")
    parser.add_argument("--count", type=int, default=10000, help="Nombre de transactions")
    parser.add_argument("--fraud-rate", type=float, default=0.03, help="Taux de fraude (0.03 = 3%)")
    parser.add_argument("--batch-size", type=int, default=500, help="Taille des batches")
    parser.add_argument("--delay", type=float, default=0.1, help="Délai entre batches (secondes)")
    args = parser.parse_args()

    # Configuration
    connection_string = os.getenv("EVENT_HUB_CONNECTION_STRING")
    eventhub_name = os.getenv("EVENT_HUB_NAME", "transactions")

    if not connection_string:
        print("❌ ERROR: EVENT_HUB_CONNECTION_STRING non défini!")
        print("   Créez un fichier .env avec:")
        print('   EVENT_HUB_CONNECTION_STRING="Endpoint=sb://..."')
        return

    print("=" * 60)
    print("   AZURE EVENT HUBS - Transaction Producer")
    print("=" * 60)
    print(f"   Transactions: {args.count}")
    print(f"   Fraud rate:   {args.fraud_rate * 100:.1f}%")
    print(f"   Batch size:   {args.batch_size}")
    print("=" * 60)

    # Initialiser
    generator = TransactionGenerator()
    producer = EventHubsProducer(connection_string, eventhub_name)
    producer.connect()

    try:
        total_sent = 0
        total_frauds = 0
        start_time = time.time()

        while total_sent < args.count:
            # Calculer la taille du prochain batch
            remaining = args.count - total_sent
            batch_size = min(args.batch_size, remaining)

            # Générer et envoyer
            transactions, fraud_count = generator.generate_batch(batch_size, args.fraud_rate)
            sent = producer.send_batch(transactions)

            total_sent += sent
            total_frauds += fraud_count

            # Afficher progression
            elapsed = time.time() - start_time
            rate = total_sent / elapsed if elapsed > 0 else 0
            progress = total_sent / args.count * 100

            print(f"📤 Envoyé: {total_sent:,}/{args.count:,} ({progress:.1f}%) | "
                  f"Fraudes: {total_frauds} | Rate: {rate:.0f} tx/s")

            # Délai entre batches
            if args.delay > 0 and total_sent < args.count:
                time.sleep(args.delay)

        # Statistiques finales
        elapsed = time.time() - start_time
        print("\n" + "=" * 60)
        print("   TERMINÉ!")
        print("=" * 60)
        print(f"   ✅ Transactions envoyées: {total_sent:,}")
        print(f"   🚨 Fraudes:               {total_frauds} ({total_frauds/total_sent*100:.2f}%)")
        print(f"   ⏱️  Temps:                 {elapsed:.2f}s")
        print(f"   📊 Débit:                 {total_sent/elapsed:.0f} tx/s")
        print("=" * 60)

    except KeyboardInterrupt:
        print("\n⚠️ Interrupted by user")
    finally:
        producer.close()


if __name__ == "__main__":
    main()

File: kafka/producer/TransactionProducer.scala
package fraud.kafka

import java.util.Properties
import org.apache.kafka.clients.producer.{KafkaProducer, ProducerConfig, ProducerRecord, Callback, RecordMetadata}
import org.apache.kafka.common.serialization.StringSerializer
import fraud.models.Transaction
import fraud.utils.ConfigManager
import scala.util.Random
import java.util.concurrent.{Executors, TimeUnit}

/**
 * TransactionProducer - Producteur Kafka de transactions
 *
 * Ce producteur simule un flux de transactions financières en temps réel.
 * Il génère des transactions normales et frauduleuses selon un ratio configurable.
 *
 * Caractéristiques:
 * - Production haute performance (10K+ messages/seconde)
 * - Transactions réalistes avec patterns de fraude
 * - Callbacks asynchrones pour confirmation
 * - Métriques de production en temps réel
 */
object TransactionProducer {

  // Données de simulation
  private val locations = Seq(
    "Paris", "London", "New York", "Tokyo", "Singapore", "Dubai",
    "Hong Kong", "Sydney", "Berlin", "Madrid", "Amsterdam", "Milan"
  )

  private val merchantCategories = Seq(
    "RETAIL", "GROCERY", "RESTAURANT", "ONLINE", "TRAVEL",
    "ENTERTAINMENT", "UTILITIES", "HEALTHCARE", "GAMBLING", "CRYPTO"
  )

  private val channels = Seq("ONLINE", "POS", "ATM", "MOBILE")
  private val currencies = Seq("EUR", "USD", "GBP", "CHF")
  private val cardTypes = Seq("CREDIT", "DEBIT")

  private val random = new Random(System.currentTimeMillis())

  /**
   * Créer une instance de KafkaProducer
   */
  def createProducer(): KafkaProducer[String, String] = {
    val props = new Properties()

    // Configuration du broker
    props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, ConfigManager.Kafka.bootstrapServers)

    // Sérialiseurs
    props.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, classOf[StringSerializer].getName)
    props.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, classOf[StringSerializer].getName)

    // Configuration de performance
    props.put(ProducerConfig.ACKS_CONFIG, ConfigManager.Kafka.producerAcks)
    props.put(ProducerConfig.RETRIES_CONFIG, ConfigManager.Kafka.producerRetries.toString)
    props.put(ProducerConfig.BATCH_SIZE_CONFIG, "16384")
    props.put(ProducerConfig.LINGER_MS_CONFIG, "5")
    props.put(ProducerConfig.BUFFER_MEMORY_CONFIG, "33554432")

    // Compression pour réduire la latence réseau
    props.put(ProducerConfig.COMPRESSION_TYPE_CONFIG, "snappy")

    // Idempotence pour exactly-once semantics
    props.put(ProducerConfig.ENABLE_IDEMPOTENCE_CONFIG, "true")

    new KafkaProducer[String, String](props)
  }

  /**
   * Générer une transaction normale
   */
  def generateNormalTransaction(customerId: String): Transaction = {
    val amount = random.nextGaussian() * 150 + 200 // Moyenne ~200, écart-type ~150
    val normalizedAmount = math.max(5, math.abs(amount)) // Minimum 5

    Transaction(
      transactionId = s"TX${System.currentTimeMillis()}-${random.nextInt(10000)}",
      customerId = customerId,
      merchantId = s"MERCH${random.nextInt(1000)}",
      amount = math.round(normalizedAmount * 100) / 100.0,
      currency = currencies(random.nextInt(currencies.length)),
      transactionType = "PURCHASE",
      channel = channels(random.nextInt(channels.length)),
      location = locations(random.nextInt(locations.length)),
      deviceId = Some(s"DEV${random.nextInt(5000)}"),
      ipAddress = Some(s"192.168.${random.nextInt(256)}.${random.nextInt(256)}"),
      timestamp = System.currentTimeMillis(),
      cardType = cardTypes(random.nextInt(cardTypes.length)),
      isInternational = random.nextDouble() < 0.1,
      merchantCategory = merchantCategories.take(8)(random.nextInt(8)), // Exclure GAMBLING, CRYPTO
      previousBalance = random.nextDouble() * 10000 + 1000,
      isFraud = Some(false)
    )
  }

  /**
   * Générer une transaction frauduleuse
   * Applique des patterns de fraude réalistes
   */
  def generateFraudulentTransaction(customerId: String): Transaction = {
    val fraudType = random.nextInt(5)

    val (amount, category, isIntl, hour) = fraudType match {
      // Type 1: Montant très élevé
      case 0 =>
        (random.nextDouble() * 15000 + 5000, "ONLINE", false, random.nextInt(24))

      // Type 2: Transaction internationale suspecte
      case 1 =>
        (random.nextDouble() * 3000 + 1000, "TRAVEL", true, random.nextInt(24))

      // Type 3: Gambling/Crypto
      case 2 =>
        val cat = if (random.nextBoolean()) "GAMBLING" else "CRYPTO"
        (random.nextDouble() * 5000 + 500, cat, random.nextBoolean(), random.nextInt(24))

      // Type 4: Transaction de nuit
      case 3 =>
        (random.nextDouble() * 2000 + 500, "ATM", false, random.nextInt(5)) // 0-5h

      // Type 5: Card testing (petits montants rapides)
      case 4 =>
        (random.nextDouble() * 10 + 0.5, "ONLINE", false, random.nextInt(24))

      case _ =>
        (random.nextDouble() * 5000 + 1000, "ONLINE", false, random.nextInt(24))
    }

    // Ajuster le timestamp pour l'heure de fraude
    val baseTimestamp = System.currentTimeMillis()
    val adjustedTimestamp = if (fraudType == 3) {
      // Mettre l'heure entre 0 et 5
      val currentHour = (baseTimestamp / 3600000) % 24
      val targetHour = random.nextInt(5)
      baseTimestamp - (currentHour - targetHour) * 3600000
    } else {
      baseTimestamp
    }

    Transaction(
      transactionId = s"TX${System.currentTimeMillis()}-${random.nextInt(10000)}",
      customerId = customerId,
      merchantId = s"MERCH${random.nextInt(1000)}",
      amount = math.round(amount * 100) / 100.0,
      currency = if (isIntl) currencies(random.nextInt(currencies.length)) else "EUR",
      transactionType = if (category == "ATM") "WITHDRAWAL" else "PURCHASE",
      channel = if (category == "ATM") "ATM" else channels(random.nextInt(channels.length)),
      location = locations(random.nextInt(locations.length)),
      deviceId = Some(s"DEV${random.nextInt(5000)}"),
      ipAddress = Some(s"${random.nextInt(256)}.${random.nextInt(256)}.${random.nextInt(256)}.${random.nextInt(256)}"),
      timestamp = adjustedTimestamp,
      cardType = cardTypes(random.nextInt(cardTypes.length)),
      isInternational = isIntl,
      merchantCategory = category,
      previousBalance = random.nextDouble() * 10000 + 1000,
      isFraud = Some(true)
    )
  }

  /**
   * Générer une transaction (normale ou frauduleuse selon le taux de fraude)
   */
  def generateTransaction(fraudRate: Double = 0.02): Transaction = {
    val customerId = s"CUST${random.nextInt(10000)}"

    if (random.nextDouble() < fraudRate) {
      generateFraudulentTransaction(customerId)
    } else {
      generateNormalTransaction(customerId)
    }
  }

  /**
   * Envoyer une transaction à Kafka
   */
  def sendTransaction(
    producer: KafkaProducer[String, String],
    topic: String,
    transaction: Transaction
  ): Unit = {
    val json = Transaction.toJson(transaction)
    val record = new ProducerRecord[String, String](topic, transaction.transactionId, json)

    producer.send(record, new Callback {
      override def onCompletion(metadata: RecordMetadata, exception: Exception): Unit = {
        if (exception != null) {
          println(s"Error sending transaction ${transaction.transactionId}: ${exception.getMessage}")
        }
      }
    })
  }

  /**
   * Démarrer la production continue de transactions
   */
  def startContinuousProduction(
    producer: KafkaProducer[String, String],
    topic: String,
    transactionsPerSecond: Int = 1000,
    fraudRate: Double = 0.02
  ): Unit = {
    val intervalNanos = 1000000000L / transactionsPerSecond
    var lastPrintTime = System.currentTimeMillis()
    var messageCount = 0L
    var fraudCount = 0L

    println(s"Starting production: $transactionsPerSecond TPS, fraud rate: ${fraudRate * 100}%")
    println("=" * 60)

    while (true) {
      val startTime = System.nanoTime()

      // Générer et envoyer une transaction
      val transaction = generateTransaction(fraudRate)
      sendTransaction(producer, topic, transaction)

      messageCount += 1
      if (transaction.isFraud.getOrElse(false)) fraudCount += 1

      // Afficher les statistiques toutes les secondes
      val currentTime = System.currentTimeMillis()
      if (currentTime - lastPrintTime >= 1000) {
        val elapsedSeconds = (currentTime - lastPrintTime) / 1000.0
        val actualTps = messageCount / elapsedSeconds
        println(f"[${java.time.LocalTime.now()}] Messages: $messageCount%,d | " +
          f"Frauds: $fraudCount%,d | TPS: $actualTps%.0f")
        messageCount = 0
        fraudCount = 0
        lastPrintTime = currentTime
      }

      // Contrôle du débit
      val elapsedNanos = System.nanoTime() - startTime
      val sleepNanos = intervalNanos - elapsedNanos
      if (sleepNanos > 0) {
        Thread.sleep(sleepNanos / 1000000, (sleepNanos % 1000000).toInt)
      }
    }
  }

  /**
   * Production en mode burst (pour tests de charge)
   */
  def burstProduction(
    producer: KafkaProducer[String, String],
    topic: String,
    messageCount: Int,
    fraudRate: Double = 0.02
  ): Long = {
    val startTime = System.currentTimeMillis()
    var fraudCount = 0

    (1 to messageCount).foreach { i =>
      val transaction = generateTransaction(fraudRate)
      sendTransaction(producer, topic, transaction)
      if (transaction.isFraud.getOrElse(false)) fraudCount += 1

      if (i % 10000 == 0) {
        println(s"Sent $i / $messageCount messages...")
      }
    }

    producer.flush()
    val endTime = System.currentTimeMillis()
    val duration = endTime - startTime
    val tps = messageCount * 1000.0 / duration

    println("=" * 60)
    println(f"Burst production completed:")
    println(f"  Total messages: $messageCount%,d")
    println(f"  Fraud messages: $fraudCount%,d (${fraudCount * 100.0 / messageCount}%.2f%%)")
    println(f"  Duration: ${duration}ms")
    println(f"  Throughput: $tps%.0f TPS")
    println("=" * 60)

    duration
  }

  /**
   * Point d'entrée principal
   */
  def main(args: Array[String]): Unit = {
    println("=" * 60)
    println("Kafka Transaction Producer")
    println("=" * 60)

    val mode = if (args.length > 0) args(0) else "continuous"
    val topic = ConfigManager.Kafka.transactionsTopic
    val producer = createProducer()

    println(s"Mode: $mode")
    println(s"Topic: $topic")
    println(s"Bootstrap Servers: ${ConfigManager.Kafka.bootstrapServers}")
    println("=" * 60)

    try {
      mode match {
        case "continuous" =>
          val tps = if (args.length > 1) args(1).toInt else 1000
          val fraudRate = if (args.length > 2) args(2).toDouble else 0.02
          startContinuousProduction(producer, topic, tps, fraudRate)

        case "burst" =>
          val count = if (args.length > 1) args(1).toInt else 100000
          val fraudRate = if (args.length > 2) args(2).toDouble else 0.02
          burstProduction(producer, topic, count, fraudRate)

        case "single" =>
          val tx = generateTransaction(0.5) // 50% chance de fraude pour test
          sendTransaction(producer, topic, tx)
          producer.flush()
          println(s"Sent single transaction: ${Transaction.toJson(tx)}")

        case _ =>
          println(s"Unknown mode: $mode")
          println("Usage: TransactionProducer [continuous|burst|single] [tps/count] [fraudRate]")
      }
    } finally {
      producer.close()
    }
  }
}
